manim-slides: Init at v5.1.3

soispha opened the pull request:
## Description of changes

This adds [Manim-slides](https://github.com/jeertmans/manim-slides/), a tool used for live presentations using manim, fixing #251914.
@dmadisetti has helped my by an incredible margin in this pr. 
## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

dmadisetti commented:
Does this not work? Why's it in draft status?

dmadisetti commented on pkgs/by-name/ma/manim-slides/package.nix:
```diff
@@ -0,0 +1,57 @@
+python.pkgs.buildPythonApplication rec {
+  pname = "manim-slides";
+  format = "pyproject";
+  version = "5.0.0";
```
up to 5.1.3 as of 24/3/20

dmadisetti commented on pkgs/by-name/ma/manim-slides/package.nix:
```diff
@@ -0,0 +1,57 @@
+    click
+    click-default-group
+    docutils
+    ipython
```
ipython is not a dep of the qt package

dmadisetti commented on pkgs/by-name/ma/manim-slides/package.nix:
```diff
@@ -0,0 +1,57 @@
+    rtoml
+    tqdm
+    # This could also be manimgl, but that is not (yet) packaged
+    manim
```
needs explicit manim source (nixpkgs or python?)
ambiguous here

dmadisetti replied:
nvm, there is only one source

dmadisetti commented on pkgs/by-name/ma/manim-slides/package.nix:
```diff
@@ -0,0 +1,57 @@
+    manim
+  ];
+
+  pythonImportsCheck = [ "manim_slides" ];
```
other tests available

dmadisetti reviewed (commented):
manim-slides can also be used a python package. notably with the jupyter magics it exposes: https://github.com/jeertmans/manim-slides/blob/a85f1c40369428b7bd948e4b809308b4ca9ee26c/docs/source/installation.md?plain=1#L111

without the qt deps

dmadisetti commented:
Opened up alternative PR using pdm and also exposing as a python package for to use ipython magics #298198 
feel free to use as reference

It works well enough. I don't want to be a maintainer of this, but I did want to tinker

soispha committed:
manim-slides: init at v5.1.3

Co-authored-by: dylan madisetti <dylan@madisetti.me>

soispha committed:
manim: relax dependencies and ignore resulting failures in tests

This change should only be temporary, as a pull request has already been
merged upstream that relaxes the dependencies [1]. Thus, we just have to
wait for a release containing this pr.

[1]: https://github.com/ManimCommunity/manim/pull/3376

Co-authored-by: dylan madisetti <dylan@madisetti.me>

dmadisetti commented on pkgs/applications/video/manim/failing_tests.nix:
```diff
@@ -64,6 +64,16 @@
+  # E           Mismatched elements: 1185[/633] / 1639680[/1639680] (0.0723[/0.0386]%)
+  # E           Max absolute difference: 125[/121]
+  # E           Max relative difference: 6.5[/1]
+  "test_ImplicitFunction"
```
Is there any indication why these a newly failing 🤔

soispha replied:
I don't really know, but I think that one of the changed dependencies had a breaking change in how they calculated tolerances. I _hope_ that we just get away with ignoring it, considering that the pr, that fixes should fix that upstream, is already merged (and we just have to wait for the next release).

soispha replied:
Maybe @friedelino might have a idea about these tests (as they are the manim maintainer).

dmadisetti replied:
Not ideal but looks like in the same vein as
   "test_Text2Color"
    "test_PointCloudDot"
    "test_Torus"
above

 I'd be OK with this as long as you leave a comment such that future maintainers can try to shrink failing_tests (e.g. Failure likely due to tolerance difference, consider removing in future updates as dependencies synchronize)

soispha replied:
Good point! I've also included a link to the pr, to give a bit of context.

dmadisetti commented on pkgs/development/python-modules/manim-slides/default.nix:
```diff
@@ -0,0 +1,93 @@
+    ffmpeg
+    # This could also be manimgl, but that is not (yet) packaged
+    manim
+  ] ++ lib.lists.optional withGui ipython;
```
`!withGui` according to the pytoml

I asked about the other deps here: https://github.com/jeertmans/manim-slides/issues/393

Could just comment the issue and move on

dmadisetti replied:
Oh OK- I just don't know how pytoml works. I think pyqt6 should be on the other branch

soispha replied:
I *think* I have the dependencies sorted out correctly now. But I'm not so sure about `qtpy`, from the pytoml file, it's just a normal dependency for every use case?

dmadisetti reviewed (commented):
Awesome! Glad I could be of help

dmadisetti reviewed (changes requested):
Looks mostly good to me

dmadisetti reviewed (commented):
Looks great, but now the order of the commits is flipped lol

collares commented:
Looks good, thanks!
